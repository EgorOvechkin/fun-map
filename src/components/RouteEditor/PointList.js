import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PointTitle from './PointTitle';
import { SortableContainer } from 'react-sortable-hoc';
import './styles/PointsList.css';

const SortableList = SortableContainer((props) =>
  <ul className='points-list'>
    {props.titles.map((title, index) => {
      return (<PointTitle key={`point-title-${index}`} index={index} title={title} deletePoint={props.deletePoint}/>);
    })}
  </ul>
);

class PointsList extends Component {
  shouldComponentUpdate(nextProps) {
    return this.props.titles.toString() !== nextProps.titles.toString();
  }

  render() {
    return (
      <SortableList
        titles={this.props.titles}
        deletePoint={this.props.deletePoint}
        onSortEnd={this.props.onSortEnd}
        distance={5}
      />
    );
  }
}

PointsList.propTypes = {
  titles: PropTypes.arrayOf(PropTypes.string),
  deletePoint: PropTypes.func,
  onSortEnd: PropTypes.func
};

export default PointsList;
