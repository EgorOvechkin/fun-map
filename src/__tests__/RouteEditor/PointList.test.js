import React from 'react';
import PointList from '../../components/RouteEditor/PointList';
import {shallow} from 'enzyme';

const generateTitles = n => Array.from({length: n}, (item, index) => `Point ${++index}`);

it('renders point list', () => {
  const titles = generateTitles(5);
  const wrapper = shallow(<PointList titles={titles} />);
  expect(wrapper).toMatchSnapshot();
});

describe('shouldComponentUpdate', () => {
  describe('when new titles is same', () => {
    it('should be false', () => {
      const oldTitles = generateTitles(3);
      const newTitles = generateTitles(3);
      const wrapper = shallow(<PointList titles={oldTitles} />);

      const shouldUpdate = wrapper.instance().shouldComponentUpdate({titles: newTitles});
      expect(shouldUpdate).toBeFalsy();
    });
  });

  describe('when new titles is different', () => {
    it('should be true', () => {
      const oldTitles = generateTitles(3);
      const newTitles = generateTitles(3).reverse();
      const wrapper = shallow(<PointList titles={oldTitles} />);

      const shouldUpdate = wrapper.instance().shouldComponentUpdate({titles: newTitles});
      expect(shouldUpdate).toBeTruthy();
    });
  });
});
