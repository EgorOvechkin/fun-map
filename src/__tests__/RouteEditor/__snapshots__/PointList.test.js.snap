// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders point list 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PointsList
    titles={
      Array [
        "Point 1",
        "Point 2",
        "Point 3",
        "Point 4",
        "Point 5",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "deletePoint": undefined,
      "distance": 5,
      "onSortEnd": undefined,
      "titles": Array [
        "Point 1",
        "Point 2",
        "Point 3",
        "Point 4",
        "Point 5",
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "deletePoint": undefined,
        "distance": 5,
        "onSortEnd": undefined,
        "titles": Array [
          "Point 1",
          "Point 2",
          "Point 3",
          "Point 4",
          "Point 5",
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
